cpp/src/replace/replace.cu

```diff
@@ -47,24 +55,41 @@ namespace{ //anonymous
    * @returns
    */
   /* ----------------------------------------------------------------------------*/
-  template <class T>
+  template <class T,
+            bool input_is_nullable, bool new_is_nullable>
   __global__
-  void replace_kernel(T*                          d_col_data,
-                      gdf_size_type                      nrows,
+  void replace_kernel(T const * __restrict__    input_data,
+                      bit_mask_t const * __restrict__ input_valid,
+                      gdf_size_type             nrows,
+                      T * __restrict__          output_data,
+                      bit_mask_t * __restrict__ output_valid,
+                      gdf_size_type*            output_null_count,
                       thrust::device_ptr<const T> old_values_begin,
                       thrust::device_ptr<const T> old_values_end,
-                      const T*                    d_new_values)
+                      const T*                    d_new_values,
+                      bit_mask_t const * __restrict__ new_valid)
   {
+
     gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
     while(i < nrows)
     {
-      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+      if ( !input_is_nullable || bit_mask::is_valid(input_valid, i)){
```

Comment:
The compiler can't optimize out this code path no matter the value of `input_is_nullable` because it is being `OR`ed with runtime dependent information (i.e., `bit_mask::is_valid(input_valid, i)`). 

`FALSE OR X`
`TRUE OR X`

Since X isn't known until runtime, the compiler can't prune the branch.

Reply:
Luckily, it's simple to change.

```suggestion
      if ( input_is_nullable and bit_mask::is_valid(input_valid, i)){
```

`FALSE and X` will always be false, and the branch can be pruned

Reply:
However, you'll have to revise the logic inside the `if` statement to make it work correctly.

Reply:
In the same way as `FALSE and X` will always be false, shouldn't `TRUE or X` be true always? 

And can we change `if ( !input_is_nullable || bit_mask::is_valid(input_valid, i))` to  
`if ( !input_is_nullable || (input_is_nullable && bit_mask::is_valid(input_valid, i)))`? I assume, it will help in branch pruning.

Reply:
> shouldn't TRUE or X be true always?

Yes, but we want to avoid the performance hit of checking `gdf_is_valid` when `nullable` is `FALSE`. By using an `OR`, the compiler can't prune away that check. So you want to setup your conditional logic in such a way that when `nullable == FALSE`, we end up with something like `if(FALSE)` such that the branch is completely removed. 

> `if ( !input_is_nullable || (input_is_nullable && bit_mask::is_valid(input_valid, i)))`

So we want to setup the conditional such that when `input_is_nullable == FALSE` that our conditional will simplify to `if(false)` and then will be pruned. So with your conditional we'd get:

` IF (not false OR false AND X)`

which simplifies to 

`IF(true OR false) -> `IF(true)`

So that won't work for avoiding the null check. 

In contrast, with the conditional I suggested:

`if ( input_is_nullable and bit_mask::is_valid(input_valid, i)){`

when `input_is_nullable == false` simplifies to 

`if( false AND X)`

which will always be `if(false)` and therefore can be pruned.